src/graphics/text.rs

```diff
@@ -271,41 +274,49 @@ impl Drawable for FontAtlas {
     }
 }
 
+const DEFAULT_TEXT_BUFFER_SIZE: usize = 64;
+
 #[derive(Debug)]
 pub struct Text {
     batch: SpriteBatch,
-    char_texture_map: HashMap<char, CharInfo>,
+    atlas: Cached<FontAtlas>,
 }
 
 impl Text {
-    pub fn new(ctx: &mut Graphics, input_text: &str, font_atlas: &FontAtlas, color: Color) -> Self {
-        let mut text = Text {
+    pub fn new(ctx: &mut Graphics, input_text: &str, font_atlas: FontAtlas, color: Color) -> Self {
+        let mut text = Self::from_cached(ctx, font_atlas.into());
+        text.set_text(input_text, color);
+        text
+    }
+
+    pub fn from_cached(ctx: &mut Graphics, mut font_atlas: Cached<FontAtlas>) -> Self {
+        let atlas = font_atlas.load_cached();
+        Text {
             batch: SpriteBatch::with_capacity(
                 ctx,
-                font_atlas.font_texture.clone(),
-                input_text.len(),
+                atlas.font_texture.clone(),
+                DEFAULT_TEXT_BUFFER_SIZE,
             ),
-            char_texture_map: font_atlas.font_map.clone(),
-        };
-        text.set_text(input_text, color);
-        text
+            atlas: font_atlas,
+        }
     }
 
     pub fn set_text(&mut self, new_text: &str, color: Color) {
+        let atlas = self.atlas.load_cached();
+        let font_map = &atlas.font_map;
         self.batch.clear();
+        self.batch.set_texture(atlas.font_texture.clone());
         let mut width: f32 = 0.;
         for c in new_text.chars() {
-            let c_info = self
-                .char_texture_map
-                .get(&c)
-                .unwrap_or(self.char_texture_map.get(&'a').unwrap());
+            let c_info = font_map.get(&c).unwrap_or(font_map.get(&'a').unwrap());
```

Comment:
We can probably pick a better default char here to fill in invalid characters than an "a", also might want to consider logging what happened to make future debugging easier

Reply:
Logging could be a very odd thing to have here, it would be better imo if we could somehow make a "bad" character or get like, the unicode question-mark-wtf-is-this glyph in. Logging could end up being super slow if this function is called on every draw; I don't think that a missing character is something we'd want to cause a game to slow to a crawl.

Reply:
If that was the case we'd need to modify the character subsets to all include the unicode question mark thing (I'm assuming you're referring to https://www.fileformat.info/info/unicode/char/fffd/index.htm)
I think for now just an ascii "?" would work, but you might want to note the unicode thing in a //TODO

Reply:
i like the idea of an ASCII `?`. we could also make it like, render two characters on top of each other or something.